woocommerce/woocommerce-gutenberg-products-block
# Store billing address to local state and persist globally to prevent loss of data

Finally done. Took longer than I wanted but I managed to fix some other issues at the same time.

1. This PR adds a local state for address data in the form of `useCheckoutAddress`. This hook persists data globally, but also stores local values so billing address is never lost.
2. If logged in, if you have a billing address separate from shipping address, the "use shipping as billing" checkbox is automatically unticked
3. I found an issue that triggered validation on all fields on change. I fixed this by using shallow equal for state/country data which are objects such as `{key: "NY", name: "New York"}`. Now validation runs on the current field only. Performance++
4. I found the issue that was causing validation to be stale once fields were unmounted. We were missing a dependency in the `useEffect` call doing the unmount 🤦 cc @Aljullu 

Fixes #2361

### How to test the changes in this Pull Request:

Smoke testing checkout, but also covering these cases:

1. Guest user. On checkout you should have the usual behaviour where "use shipping as billing" is ticked and billing is blank.
2. Logged in user. Edit your profile and use a different billing address. On checkout, "use shipping as billing" is unchecked, and both addresses are populated in the form.
3. Test inputting address and confirming validation is working and values are persisted.
